sklearn/preprocessing/_discretization.py

```diff
@@ -152,6 +153,15 @@ class KBinsDiscretizer(TransformerMixin, BaseEstimator):
            [ 0.5,  3.5, -1.5,  1.5]])
     """
 
+    _parameter_constraints = {
+        "nbins": [Interval(numbers.Integral, 2, None, closed="left"), "array-like"],
+        "encode": [StrOptions({"onehot", "onehot-dense", "ordinal"})],
+        "strategy": [StrOptions({"uniform", "quantile", "kmeans"})],
+        "dtype": [numbers.Real, None],
```

Comment:
Currently, it's not possible to specify that a parameter can be a given type. Let's put ``type`` for now.
```suggestion
        "dtype": [type, None],  # TODO: TypeOptions constraint
```